### PR TITLE
In JS, class supports `@template` tag for declaring type parameters

### LINKED ISSUE #18135: Unable to make class generic with JSDoc
**TypeScript Version:**  2.6.0-dev.20170829

```js
/**
 * @template T
 */
class MyArray {
  constructor() {
    /** @type {Array<T>} */
    this.array = [];
  }
}
/** @type {MyArray<string>} */
let x = new MyArray();
```

**Expected behavior:**
`x` should contain an array of strings.

**Actual behavior:**
`[js] Cannot find name 'T'.`
`[js] Type 'MyArray' is not generic.`

### LINKED ISSUE #23384: @extends/@augments should work with @template
```ts
// @ts-check

/**
 * @template {T}
 * @extends {Set<T>}
 */
export class SortableSet extends Set {

}
```

Current behavior is

1. `Cannot find name 'T'.`
2. `Base constructors must all have the same return type.`

@TheLarkInn

### LINKED ISSUE #23385: @template tag doesn't work in @typedefs
```ts
/**
 * @template {T}
 * @typedef {(x: T, y: T) => number} SetComparer
 */
```

**Expected**: `x` and `y` have type `T`
**Actual**: `x` any `y` have type `any`, and `T` is associated with whatever declaration follows the JSDoc comment...

For example the following

```ts
/**
 * @template {T}
 * @typedef {(x: T, y: T) => number} SetComparer
 */
class C {
}
```

makes `C` generic on `T`.

### LINKED ISSUE #23459: @template tag is not visible in class body
Found playing with https://github.com/lukeed/tmp-cache

```ts
// @ts-check

/**
 * @template {T}
 */
class Foo {
    /**
     * 
     * @param {T} x 
     */
    foo(x) {	
    }
}
```

This will give the following error on `T` in the `@param` tag.

```
Cannot find name 'T'.
```

### PR DESCRIPTION
**Pull request by sandersn**
Previously, classes did not understand the `@template` tag, so there was no way to declare and use type parameters on classes in Javascript.

This change makes the checker get the "effective" type parameters for a class, which means that Javascript classes also check their jsdoc for `@template` tags. The rest of the type parameter machinery is unchanged.

Edit: this PR now makes type parameters work with constructor functions as well. This required changes in name resolution and class instantiation since type parameters can be looked up outside their scope, and the anonymous type created by newing a Javascript constructor can now be instantiated.

Only the class support is needed to unblock webpack/webpack#7046 and webpack/webpack#7050, however.

Fixes #23459
Fixes #23384  
Fixes #23385
Fixes #18135

### CODE REVIEW

**Comment by sandersn** [comment]
CI failed because I missed a fourslash test. It's updated now.

**Comment by sandersn** [comment]
I added support for `@template` in Javascript constructors as well. This required changes in name resolution and class instantiation since type parameters can be looked up outside their scope, and the anonymous type created by newing a Javascript constructor can now be instantiated.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2060,10 +2061,10 @@ namespace ts {
             let symbol: Symbol;
             if (name.kind === SyntaxKind.Identifier) {
                 const message = meaning === namespaceMeaning ? Diagnostics.Cannot_find_namespace_0 : Diagnostics.Cannot_find_name_0;
-
-                symbol = resolveName(location || name, name.escapedText, meaning, ignoreErrors ? undefined : message, name, /*isUse*/ true);
+                const secondarySymbol = resolveEntityNameFromJSPrototype(name, meaning);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: consider adding `isInJavascriptFile` just for clarity.

Comment (sandersn) [reply src/compiler/checker.ts]: And maybe a comment that says "secondarySymbol is undefined except for type references in jsdoc on javascript prototype assignments"

Or something simpler than that. Hopefully.

**Pull request merged**